catalyst-toolbox-lib/src/recovery/tally.rs

```diff
@@ -0,0 +1,502 @@
+use jormungandr_lib::{
+    interfaces::{
+        Address, Block0Configuration, FragmentLogDeserializeError, Initial, PersistentFragmentLog,
+        SlotDuration as Block0SlotDuration,
+    },
+    time::SecondsSinceUnixEpoch,
+};
+use jormungandr_testing_utils::wallet::Wallet;
+use std::collections::{HashMap, HashSet};
+
+#[allow(clippy::large_enum_variant)]
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    DeserializeError(#[from] jormungandr_lib::interfaces::FragmentLogDeserializeError),
+
+    #[error(transparent)]
+    LedgerError(#[from] chain_impl_mockchain::ledger::Error),
+
+    #[error("Couldn't initiate a new wallet")]
+    WalletError(#[from] jormungandr_testing_utils::wallet::WalletError),
+
+    #[error(transparent)]
+    Block0ConfigurationError(#[from] jormungandr_lib::interfaces::Block0ConfigurationError),
+
+    #[error("Block0 do not contain any voteplan")]
+    MissingVoteplanError,
+
+    #[error("Could not verify transaction {id} signature with range {range:?}")]
+    InvalidTransactionSignature {
+        id: String,
+        range: std::ops::Range<u32>,
+    },
+
+    #[error(
+        "Fragment with id {id} and spending counter value {spending_counter} was already processed"
+    )]
+    DuplicatedFragment { id: String, spending_counter: u32 },
+
+    #[error("Tried to vote with a non registered account: {0}")]
+    NonVotingAccount(String),
+}
+
+fn timestamp_to_system_time(ts: SecondsSinceUnixEpoch) -> SystemTime {
+    SystemTime::UNIX_EPOCH.add(Duration::new(ts.to_secs(), 0))
+}
+
+fn fragment_log_timestamp_to_blockdate(
+    timestamp: SecondsSinceUnixEpoch,
+    timeframe: &TimeFrame,
+    ledger: &Ledger,
+) -> Option<BlockDate> {
+    let slot = timestamp_to_system_time(timestamp);
+    let new_slot = timeframe.slot_at(&slot)?;
+    let epoch_position = ledger.era().from_slot_to_era(new_slot)?;
+    Some(BlockDate::from(epoch_position))
+}
+
+fn timeframe_from_block0_start_and_slot_duration(
+    block0_start: SecondsSinceUnixEpoch,
+    slot_duration: Block0SlotDuration,
+) -> TimeFrame {
+    let timeline = Timeline::new(timestamp_to_system_time(block0_start));
+
+    TimeFrame::new(
+        timeline,
+        SlotDuration::from_secs(<u8>::from(slot_duration) as u32),
+    )
+}
+
+fn committee_id_to_address(id: CommitteeIdDef) -> Address {
+    let id = CommitteeId::from(id);
+    let pk = id.public_key();
+    chain_addr::Address(Discrimination::Production, Kind::Account(pk)).into()
+}
+
+fn voteplans_from_block0(block0: &Block) -> HashMap<VotePlanId, VotePlan> {
+    block0
+        .fragments()
+        .filter_map(|fragment| {
+            if let Fragment::VotePlan(tx) = fragment {
+                let voteplan = tx.as_slice().payload().into_payload();
+                Some((voteplan.to_id(), voteplan))
+            } else {
+                None
+            }
+        })
+        .collect()
+}
+
+fn verify_original_tx(
```

Comment:
What is the motivation to do that? An heuristic to check of the fragment should be rejected unconditionally?

Reply:
It is just to check the signature of the content, but without checking the actual spending counter (and instead check a possible range of them).